saw-remote-api/src/SAWServer/JVMCrucibleSetup.hs

```diff
@@ -0,0 +1,197 @@
+{-# LANGUAGE DeriveFoldable #-}
+{-# LANGUAGE DeriveFunctor #-}
+{-# LANGUAGE DeriveTraversable #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE ImplicitParams #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE PartialTypeSignatures #-}
+module SAWServer.JVMCrucibleSetup
+  ( startJVMSetup
+  , jvmLoadClass
+  , compileJVMContract
+  ) where
+
+import Control.Lens
+import Control.Monad.IO.Class
+import Control.Monad.State
+import Data.Aeson (FromJSON(..), withObject, (.:))
+import Data.ByteString (ByteString)
+import Data.Foldable
+import Data.Map (Map)
+import qualified Data.Map as Map
+import Data.Maybe
+import qualified Data.Text as T
+
+import qualified Cryptol.Parser.AST as P
+import Cryptol.Utils.Ident (mkIdent)
+import qualified Lang.Crucible.JVM as CJ
+import qualified Language.JVM.Common as JVM
+import SAWScript.Crucible.Common.MethodSpec as MS (SetupValue(..))
+import SAWScript.Crucible.JVM.Builtins
+import SAWScript.Crucible.JVM.BuiltinsJVM
+import qualified SAWScript.Crucible.JVM.MethodSpecIR as CMS
+import SAWScript.JavaExpr (JavaType(..))
+import SAWScript.Options (defaultOptions)
+import SAWScript.Value (BuiltinContext, JVMSetupM(..), biSharedContext)
+import qualified Verifier.SAW.CryptolEnv as CEnv
+import Verifier.SAW.CryptolEnv (CryptolEnv)
+import Verifier.SAW.TypedTerm (TypedTerm)
+
+import Argo
+import SAWServer
+import SAWServer.Data.Contract
+import SAWServer.Data.JVMType
+import SAWServer.Data.SetupValue ()
+import SAWServer.CryptolExpression (getTypedTermOfCExp)
+import SAWServer.Exceptions
+import SAWServer.OK
+import SAWServer.TopLevel
+import SAWServer.TrackFile
+
+startJVMSetup :: StartJVMSetupParams -> Method SAWState OK
+startJVMSetup (StartJVMSetupParams n) =
+  do pushTask (JVMSetup n [])
+     ok
+
+data StartJVMSetupParams
+  = StartJVMSetupParams ServerName
+
+instance FromJSON StartJVMSetupParams where
+  parseJSON =
+    withObject "params for \"SAW/Crucible setup\"" $ \o ->
+    StartJVMSetupParams <$> o .: "name"
+
+data ServerSetupVal = Val (SetupValue CJ.JVM)
+
+-- TODO: this is an extra layer of indirection that could be collapsed, but is easy to implement for now.
+compileJVMContract ::
+  (FilePath -> IO ByteString) ->
+  BuiltinContext ->
+  CryptolEnv ->
+  Contract JavaType (P.Expr P.PName) ->
+  JVMSetupM ()
+compileJVMContract fileReader bic cenv c = interpretJVMSetup fileReader bic cenv (reverse steps)
```

Comment:
Why reverse the steps here, and then again later?

Reply:
Oh, good catch. This is copied from the LLVM interface, so it probably needs to be changed there, too.

Reply:
:cold_sweat: